### PR TITLE
Major refactoring: modular player

### LINKED ISSUE #192: Unable to connect to subsonic server
I have been using a local copy of Jamstash 3.3 for a long time and it has worked fine. But when I now try and upgrade to the newest version 4.3 I get an error as soon as I save my setting "Unable to connect to Subsonic Server". I can connect to the demo server. My server uses a different port - 4040 -   but the demo server has no port specified. Maybe that can be an issue.

I have also tried the Chrome App. I uninstalled it and installe it again. I get the same error in the Chrome app.

I have tried to enable "debug mode" but I cannot see any debug info.

I am running Chrome on Windows 8.1

### PR DESCRIPTION
**Pull request by Hyzual**
Hi again !

Second giant PR. I wanted to separate this from the Grunt & Karma things, so that it's easier to understand and merge. I kinda wish I could have made it smaller, but I didn't want to submit a PR with half-working things...

**TL;DR** : I rewrote the player part from scratch in order to make it more modular and therefore make it easier for us to add or remove functionalities.

See the result on my server : http://jamstash.furinax.com

It's mostly a tech version, on the user-side a few bugs should be fixed but there are no new features.
The PR fixes different problems :

## Simplifying error-handling when making requests to Subsonic

- Adds a generic `subsonicRequest()` that handles protocols and errors. It makes `getRandomSongs`, `getStarredSongs`, `scrobble`, etc. very light : they only deal with getting songs (or scrobbling).
- Fixes #192 . While I was writing this "wrapper" to make it easier for me to deal with protocols and errors, I noticed that `ping()` didn't use `$http.JSONP()`. That's probably why it didn't work with protocol JSONP : you have to use that method so that angular can create a callback for Subsonic to use. Anyway, now it should work with any protocol and display errors such as a bad username or missing password to the user.

## Separation of concerns :  a better architecture through modularity

I had taken a look at issue #184, and I found it crazily difficult to untangle player-service.js. Everything was mixed in the player service : we used jquery, jplayer, manipulated DOM, called other services, made HTTP requests. We need to keep these separate so we don't end up with 50 dependencies and can add or remove functionnalities more easily.

In order to do this, I did the following :

- New rewritten player service. It manages the queue and no longer depends on the rootScope. If someone wants to do anything like play a song or fill the queue, they can use the player service. The player service does not do anything HTTP-related. Scrobbling songs is delegated to the subsonic service, since we use subsonic to scrobble for us.
- Added a separate controller for the queue. It uses the player service and adds display-related things.
- Added a separate controller for the player controls. it uses the player service.
- Added a separate directive that controls jplayer. We should only manipulate DOM and use jplayer there. We never have to even know jplayer anywhere else. This will make it a lot easier to write a complete jukebox mode : the player service behaves the same way, but talks to a different directive.
- Added a persistence service. its job is to store the playing queue and current song in localStorage. I also added `angular-locker` as a dependency. It performs the checks and error-handling for us, so we can focus on feeding it info and retrieving it.
- Mostly rewritten notifications service. `showNotification()` now takes a song as parameter and it builds the notification itself, so the player-directive doesn't have to depend on utils.
- Added a Page service. It controls the tab title in order to display (or scroll) a song's title. The directive doesn't have to care how the title works, it justs tells it to display a song.
- Removed many things from main-controller, utils, settings, etc. They are all moved somewhere else where it makes more sense
- Added unit-tests for everything above. We currently stand at about 30% of the entire JS code unit-tested. The player controllers, directive, service, etc. are all about 80% unit-tested. Some of the remaining code does not need to be tested (getters, random, etc.). Unit-tests will make it incredibly easier to change or remove things without breaking the rest. If you use `grunt serve`, thanks to Karma you'll always know if something player-related stops working.
- Updated bower dependencies: jplayer, jquery and jquery-ui

It fixes various bugs :
- The "playing" icon sometimes did not disappear from songs, so we ended up having more than one song marked as "playing". I use a different system where instead of modifying the song, I store the current song in the player. Since there can be only one, this bug won't happen.
- The starred auto playlist now correctly appears "selected" when clicked. It was something I had forgotten to modify.

### Functionalities

Here are the functionalities I rewrote and tested : 
- play using the service
- play / pause in the directive
- mute / unmute
- previous track
- next track
- shuffle queue
- empty queue
- load a song but do not play it
- play a song after loading it
- add songs to the queue
- remove songs from the queue
- scrobble a song
- repeat a song
- repeat the playlist
- auto-play playlist : using subsonic to fill the queue with random songs
- load the queue from localStorage
- load track at given position from localStorage
- save queue to localStorage
- save track position to localStorage
- delete queue from localStorage
- delete track from localStorage
- replay if error in jPlayer
- scroll queue to playing song (in queue controller)
- notify the playing song
- play different formats (oga, m4a)
- display the current song as playing in the queue
- scroll the song's title in the tab bar
- open the next album art in fancybox when changing song
- scrobble again when the song is restarted
- play a song in jukebox mode
- pause a song in jukebox mode

Everything should work the way it did before. I spent a lot of time testing everything, but it's always possible that I missed something. I have removed some debug code too, like writing the song specs to HTML near jplayer. I don't really know if it was useful, but in any case let me know if I removed something important.

Take all the time you need testing this. I'll keep using it on my server too.
Best regards !

### CODE REVIEW

**Comment by tsquillario** [comment]
Yea organizing the player has always been a struggle for me. The circular dependencies gave me fits. This has been long overdue! I'll push it to http://beta.jamstash.com for additional testing. 

I had a look through the the commits, good stuff. No need for the pull request write up, your git commit comments do the job.

**Comment by Hyzual** [comment]
Thanks :) Okay, next time I'll write less in the PR, it'll probably save us both some time.

**Comment by tsquillario** [comment]
The only thing I've found is the play/pause when I hit Spacebar. It's calling player.playPauseSong()

What's the right way to get the status from jPlayer?

**Comment by Hyzual** [comment]
Ah, yes, I forgot about the key shortcuts. I think the best way to do it
would be in the directive, so you can get access to jPlayer.

**Comment by nibzo1** [comment]
Great works guys :)

Although i have noticed the spacebar issue too
